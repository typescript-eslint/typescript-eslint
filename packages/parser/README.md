<h1 align="center">TypeScript ESLint Parser</h1>

<p align="center">An ESLint custom parser which leverages <a href="https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/typescript-estree">TypeScript ESTree</a> to allow for ESLint to lint TypeScript source code.</p>

<p align="center">
    <a href="https://dev.azure.com/typescript-eslint/TypeScript%20ESLint/_build/latest?definitionId=1&branchName=master"><img src="https://img.shields.io/azure-devops/build/typescript-eslint/TypeScript%20ESLint/1/master.svg?label=%F0%9F%9A%80%20Azure%20Pipelines&style=flat-square" alt="Azure Pipelines"/></a>
    <a href="https://github.com/typescript-eslint/typescript-eslint/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/typescript-estree.svg?style=flat-square" alt="GitHub license" /></a>
    <a href="https://www.npmjs.com/package/@typescript-eslint/parser"><img src="https://img.shields.io/npm/v/@typescript-eslint/parser.svg?style=flat-square" alt="NPM Version" /></a>
    <a href="https://www.npmjs.com/package/@typescript-eslint/parser"><img src="https://img.shields.io/npm/dm/@typescript-eslint/parser.svg?style=flat-square" alt="NPM Downloads" /></a>
    <a href="http://commitizen.github.io/cz-cli/"><img src="https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?style=flat-square" alt="Commitizen friendly" /></a>
</p>

## Installation:

```sh
npm install @typescript-eslint/parser --save-dev
```

## Usage

In your ESLint configuration file, set the `parser` property:

```json
{
  "parser": "@typescript-eslint/parser"
}
```

There is sometimes an incorrect assumption that the parser itself is what does everything necessary to facilitate the use of ESLint with TypeScript. In actuality, it is the combination of the parser _and_ one or more plugins which allow you to maximize your usage of ESLint with TypeScript.

For example, once this parser successfully produces an AST for the TypeScript source code, it might well contain some information which simply does not exist in a standard JavaScript context, such as the data for a TypeScript-specific construct, like an `interface`.

The core rules built into ESLint, such as `indent` have no knowledge of such constructs, so it is impossible to expect them to work out of the box with them.

Instead, you also need to make use of one more plugins which will add or extend rules with TypeScript-specific features.

By far the most common case will be installing the [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin) plugin, but there are also other relevant options available such a [@typescript-eslint/eslint-plugin-tslint](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin-tslint).

## Configuration

The following additional configuration options are available by specifying them in [`parserOptions`](https://eslint.org/docs/user-guide/configuring#specifying-parser-options) in your ESLint configuration file.

- **`ecmaFeatures.jsx`** - default `false`. Enable parsing JSX when `true`. More details can be found [here](https://www.typescriptlang.org/docs/handbook/jsx.html).

  NOTE: this setting does not effect known file types (.js, .jsx, .ts, .tsx, .json) because the typescript compiler has its own internal handling for known file extensions. The exact behaviour is as follows:

  - if `parserOptions.project` is _not_ provided:
    - `.js`, `.jsx`, `.tsx` files are parsed as if this is true.
    - `.ts` files are parsed as if this is false.
    - unknown extensions (`.md`, `.vue`) will respect this setting.
  - if `parserOptions.project` is provided (i.e. you are using rules with type information):
    - `.js`, `.jsx`, `.tsx` files are parsed as if this is true.
    - `.ts` files are parsed as if this is false.
    - "unknown" extensions (`.md`, `.vue`) **are parsed as if this is false**.

- **`useJSXTextNode`** - default `true`. Please set `false` if you use this parser on ESLint v4. If this is `false`, the parser creates the AST of JSX texts as the legacy style.

- **`project`** - default `undefined`. This option allows you to provide a path to your project's `tsconfig.json`. **This setting is required if you want to use rules which require type information**. You may want to use this setting in tandem with the `tsconfigRootDir` option below.

  - Accepted values:

    ```js
    // path
    project: './tsconfig.json';

    // glob pattern
    project: './packages/**/tsconfig.json';

    // array of paths and/or glob patterns
    project: [
      './packages/**/tsconfig.json',
      './separate-package/tsconfig.json',
    ];
    ```

  - If you use project references, TypeScript will not automatically use project references to resolve files. This means that you will have to add each referenced tsconfig to the `project` field either separately, or via a glob.

  - TypeScript will ignore files with duplicate filenames in the same folder (for example, `src/file.ts` and `src/file.js`). TypeScript purposely ignore all but one of the files, only keeping the one file with the highest priority extension (the extension priority order (from highest to lowest) is `.ts`, `.tsx`, `.d.ts`, `.js`, `.jsx`). For more info see #955.

  - Note that if this setting is specified and `createDefaultProgram` is not, you must only lint files that are included in the projects as defined by the provided `tsconfig.json` files. If your existing configuration does not include all of the files you would like to lint, you can create a separate `tsconfig.eslint.json` as follows:

    ```jsonc
    {
      // extend your base config so you don't have to redefine your compilerOptions
      "extends": "./tsconfig.json",
      "include": [
        "src/**/*.ts",
        "test/**/*.ts",
        "typings/**/*.ts",
        // etc

        // if you have a mixed JS/TS codebase, don't forget to include your JS files
        "src/**/*.js"
      ]
    }
    ```

- **`tsconfigRootDir`** - default `undefined`. This option allows you to provide the root directory for relative tsconfig paths specified in the `project` option above.

- **`extraFileExtensions`** - default `undefined`. This option allows you to provide one or more additional file extensions which should be considered in the TypeScript Program compilation. E.g. a `.vue` file

- **`warnOnUnsupportedTypeScriptVersion`** - default `true`. This option allows you to toggle the warning that the parser will give you if you use a version of TypeScript which is not explicitly supported

- **`createDefaultProgram`** - default `false`. This option allows you to request that when the `project` setting is specified, files will be allowed when not included in the projects defined by the provided `tsconfig.json` files. **Using this option will incur significant performance costs. This option is primarily included for backwards-compatibility.** See the **`project`** section above for more information.

### .eslintrc.json

```json
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    },
    "useJSXTextNode": true,
    "project": "./tsconfig.json",
    "tsconfigRootDir": "../../",
    "extraFileExtensions": [".vue"]
  }
}
```

## Supported TypeScript Version

Please see https://github.com/typescript-eslint/typescript-eslint for the supported TypeScript version.

**Please ensure that you are using a supported version before submitting any issues/bug reports.**

## Reporting Issues

Please use the @typescript-eslint/parser issue template when creating your issue and fill out the information requested as best you can. This will really help us when looking into your issue.

## License

TypeScript ESLint Parser is licensed under a permissive BSD 2-clause license.
