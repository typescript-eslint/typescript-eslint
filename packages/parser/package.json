{
  "name": "@typescript-eslint/parser",
  "version": "2.21.0",
  "description": "An ESLint custom parser which leverages TypeScript ESTree",
  "main": "dist/parser.js",
  "types": "dist/parser.d.ts",
  "files": [
    "dist",
    "README.md",
    "LICENSE"
  ],
  "engines": {
    "node": "^8.10.0 || ^10.13.0 || >=11.10.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/typescript-eslint/typescript-eslint.git",
    "directory": "packages/parser"
  },
  "bugs": {
    "url": "https://github.com/typescript-eslint/typescript-eslint/issues"
  },
  "license": "BSD-2-Clause",
  "keywords": [
    "ast",
    "ecmascript",
    "javascript",
    "typescript",
    "parser",
    "syntax",
    "eslint"
  ],
  "scripts": {
    "build": "tsc -b tsconfig.build.json",
    "clean": "tsc -b tsconfig.build.json --clean",
    "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
    "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
    "test": "jest --coverage",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "peerDependencies": {
    "eslint": "^5.0.0 || ^6.0.0"
  },
  "dependencies": {
    "@types/eslint-visitor-keys": "^1.0.0",
    "@typescript-eslint/experimental-utils": "2.21.0",
    "@typescript-eslint/typescript-estree": "2.21.0",
    "eslint-visitor-keys": "^1.1.0"
  },
  "devDependencies": {
    "@types/glob": "^7.1.1",
    "@typescript-eslint/shared-fixtures": "2.21.0",
    "glob": "*"
  },
  "peerDependenciesMeta": {
    "typescript": {
      "optional": true
    }
  },
  "funding": {
    "type": "opencollective",
    "url": "https://opencollective.com/typescript-eslint"
  }
}
