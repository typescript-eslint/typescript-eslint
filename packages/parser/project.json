{
  "name": "parser",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "type": "library",
  "implicitDependencies": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/parser/**/*.ts"]
      }
    }
  }
}
