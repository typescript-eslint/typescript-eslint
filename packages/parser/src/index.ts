export { parse, parseForESLint, ParserOptions } from './parser';
export { ParserServices } from '@typescript-eslint/typescript-estree';
export { clearCaches } from '@typescript-eslint/typescript-estree';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
