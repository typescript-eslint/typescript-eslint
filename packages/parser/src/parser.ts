import { ParserOptions, TSESTree, Lib } from '@typescript-eslint/types';
import {
  parseAndGenerateServices,
  ParserServices,
  TSESTreeOptions,
  visitorKeys,
} from '@typescript-eslint/typescript-estree';
import { analyze, ScopeManager } from '@typescript-eslint/scope-manager';
import debug from 'debug';
import { ScriptTarget } from 'typescript';

const log = debug('typescript-eslint:parser:parser');

interface ParseForESLintResult {
  ast: TSESTree.Program & {
    range?: [number, number];
    tokens?: TSESTree.Token[];
    comments?: TSESTree.Comment[];
  };
  services: ParserServices;
  visitorKeys: typeof visitorKeys;
  scopeManager: ScopeManager;
}

function validateBoolean(
  value: boolean | undefined,
  fallback = false,
): boolean {
  if (typeof value !== 'boolean') {
    return fallback;
  }
  return value;
}

const LIB_FILENAME_REGEX = /lib\.(.+)\.d\.ts/;
function getLib(services: ParserServices): Lib[] {
  const compilerOptions = services.program.getCompilerOptions();
  if (compilerOptions.lib) {
    return compilerOptions.lib
      .map(lib => {
        const match = LIB_FILENAME_REGEX.exec(lib.toLowerCase());
        if (!match) {
          return null;
        }

        return match[1] as Lib;
      })
      .filter(l => l != null) as Lib[];
  }

  const target = compilerOptions.target ?? ScriptTarget.ES5;
  // https://github.com/Microsoft/TypeScript/blob/59ad375234dc2efe38d8ee0ba58414474c1d5169/src/compiler/utilitiesPublic.ts#L13-L32
  switch (target) {
    case ScriptTarget.ESNext:
      return ['esnext.full'];
    case ScriptTarget.ES2020:
      return ['es2020.full'];
    case ScriptTarget.ES2019:
      return ['es2019.full'];
    case ScriptTarget.ES2018:
      return ['es2018.full'];
    case ScriptTarget.ES2017:
      return ['es2017.full'];
    case ScriptTarget.ES2016:
      return ['es2016.full'];
    case ScriptTarget.ES2015:
      return ['es6'];
    default:
      return ['lib'];
  }
}

function parse(
  code: string,
  options?: ParserOptions,
): ParseForESLintResult['ast'] {
  return parseForESLint(code, options).ast;
}

function parseForESLint(
  code: string,
  options?: ParserOptions | null,
): ParseForESLintResult {
  if (!options || typeof options !== 'object') {
    options = {};
  } else {
    options = { ...options };
  }
  // https://eslint.org/docs/user-guide/configuring#specifying-parser-options
  // if sourceType is not provided by default eslint expect that it will be set to "script"
  if (options.sourceType !== 'module' && options.sourceType !== 'script') {
    options.sourceType = 'script';
  }
  if (typeof options.ecmaFeatures !== 'object') {
    options.ecmaFeatures = {};
  }

  const parserOptions: TSESTreeOptions = {};
  Object.assign(parserOptions, options, {
    useJSXTextNode: validateBoolean(options.useJSXTextNode, true),
    jsx: validateBoolean(options.ecmaFeatures.jsx),
  });

  if (typeof options.filePath === 'string') {
    const tsx = options.filePath.endsWith('.tsx');
    if (tsx || options.filePath.endsWith('.ts')) {
      parserOptions.jsx = tsx;
    }
  }

  /**
   * Allow the user to suppress the warning from typescript-estree if they are using an unsupported
   * version of TypeScript
   */
  const warnOnUnsupportedTypeScriptVersion = validateBoolean(
    options.warnOnUnsupportedTypeScriptVersion,
    true,
  );
  if (!warnOnUnsupportedTypeScriptVersion) {
    parserOptions.loggerFn = false;
  }

  const { ast, services } = parseAndGenerateServices(code, parserOptions);
  ast.sourceType = options.sourceType;

  // automatically apply the libs configured for the program
  if (services.hasFullTypeInformation && options.lib == null) {
    options.lib = getLib(services);
    log('Resolved libs from program: %o', options.lib);
  }

  const analyzeOptions = {
    ecmaVersion: options.ecmaVersion,
    globalReturn: options.ecmaFeatures.globalReturn,
    lib: options.lib,
    sourceType: options.sourceType,
  };
  const scopeManager = analyze(ast, analyzeOptions);

  return { ast, services, scopeManager, visitorKeys };
}

export { parse, parseForESLint, ParserOptions };
