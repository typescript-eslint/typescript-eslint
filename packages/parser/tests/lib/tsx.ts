import { Linter } from 'eslint';
import fs from 'fs';
import glob from 'glob';
import * as parser from '../../src/parser';
import {
  createScopeSnapshotTestBlock,
  formatSnapshotName,
} from '../tools/test-utils';

const FIXTURES_DIR =
  '../../node_modules/@typescript-eslint/shared-fixtures/fixtures/tsx';
const testFiles = glob.sync(`${FIXTURES_DIR}/**/*.src.tsx`);

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe('TSX', () => {
  testFiles.forEach(filename => {
    const code = fs.readFileSync(filename, 'utf8');
    const config = {
      useJSXTextNode: true,
      ecmaFeatures: {
        jsx: true,
      },
    };
    it(
      formatSnapshotName(filename, FIXTURES_DIR, '.tsx'),
      createScopeSnapshotTestBlock(code, config),
    );
  });

  describe("if the filename ends with '.tsx', enable jsx option automatically.", () => {
    const linter = new Linter();
    linter.defineParser('@typescript-eslint/parser', parser);

    it('filePath was not provided', () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
      };
      const messages = linter.verify(code, config);

      expect(messages).toStrictEqual([
        {
          column: 18,
          fatal: true,
          line: 1,
          message: "Parsing error: '>' expected.",
          ruleId: null,
          severity: 2,
        },
      ]);
    });

    it("filePath was not provided and 'jsx:true' option", () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
        parserOptions: {
          ecmaFeatures: {
            jsx: true,
          },
        },
      };
      const messages = linter.verify(code, config);

      expect(messages).toStrictEqual([]);
    });

    it('test.ts', () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
      };
      const messages = linter.verify(code, config, { filename: 'test.ts' });

      expect(messages).toStrictEqual([
        {
          column: 18,
          fatal: true,
          line: 1,
          message: "Parsing error: '>' expected.",
          ruleId: null,
          severity: 2,
        },
      ]);
    });

    it("test.ts with 'jsx:true' option", () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
        parserOptions: {
          ecmaFeatures: {
            jsx: true,
          },
        },
      };
      const messages = linter.verify(code, config, { filename: 'test.ts' });

      expect(messages).toStrictEqual([
        {
          column: 18,
          fatal: true,
          line: 1,
          message: "Parsing error: '>' expected.",
          ruleId: null,
          severity: 2,
        },
      ]);
    });

    it('test.tsx', () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
      };
      const messages = linter.verify(code, config, { filename: 'test.tsx' });

      expect(messages).toStrictEqual([]);
    });

    it("test.tsx with 'jsx:false' option", () => {
      const code = 'const element = <T/>';
      const config = {
        parser: '@typescript-eslint/parser',
        parserOptions: {
          ecmaFeatures: {
            jsx: false,
          },
        },
      };
      const messages = linter.verify(code, config, { filename: 'test.tsx' });

      expect(messages).toStrictEqual([]);
    });
  });
});
