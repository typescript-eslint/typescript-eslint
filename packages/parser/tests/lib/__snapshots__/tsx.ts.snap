// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSX fixtures/generic-jsx-element.src 1`] = `
Object {
  "$id": 1,
  "block": Object {
    "range": Array [
      0,
      34,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 0,
      "block": Object {
        "range": Array [
          0,
          34,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 1,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 0,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 1,
  },
  "variables": Array [],
}
`;

exports[`TSX fixtures/generic-jsx-opening-element.src 1`] = `
Object {
  "$id": 1,
  "block": Object {
    "range": Array [
      0,
      46,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 0,
      "block": Object {
        "range": Array [
          0,
          46,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 1,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 0,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 1,
  },
  "variables": Array [],
}
`;

exports[`TSX fixtures/react-typed-props.src 1`] = `
Object {
  "$id": 7,
  "block": Object {
    "range": Array [
      0,
      164,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 6,
      "block": Object {
        "range": Array [
          0,
          164,
        ],
        "type": "Program",
      },
      "childScopes": Array [
        Object {
          "$id": 5,
          "block": Object {
            "range": Array [
              80,
              164,
            ],
            "type": "FunctionDeclaration",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [
            Object {
              "$id": 4,
              "from": Object {
                "$ref": 5,
              },
              "identifier": Object {
                "name": "props",
                "range": Array [
                  136,
                  141,
                ],
                "type": "Identifier",
              },
              "kind": "r",
              "resolved": Object {
                "$ref": 3,
              },
              "writeExpr": undefined,
            },
          ],
          "throughReferences": Array [],
          "type": "function",
          "upperScope": Object {
            "$ref": 6,
          },
          "variableMap": Object {
            "arguments": Object {
              "$ref": 2,
            },
            "props": Object {
              "$ref": 3,
            },
          },
          "variableScope": Object {
            "$ref": 5,
          },
          "variables": Array [
            Object {
              "$id": 2,
              "defs": Array [],
              "eslintUsed": undefined,
              "identifiers": Array [],
              "name": "arguments",
              "references": Array [],
              "scope": Object {
                "$ref": 5,
              },
            },
            Object {
              "$id": 3,
              "defs": Array [
                Object {
                  "name": Object {
                    "name": "props",
                    "range": Array [
                      93,
                      105,
                    ],
                    "type": "Identifier",
                  },
                  "node": Object {
                    "range": Array [
                      80,
                      164,
                    ],
                    "type": "FunctionDeclaration",
                  },
                  "parent": null,
                  "type": "Parameter",
                },
              ],
              "eslintUsed": undefined,
              "identifiers": Array [
                Object {
                  "name": "props",
                  "range": Array [
                    93,
                    105,
                  ],
                  "type": "Identifier",
                },
              ],
              "name": "props",
              "references": Array [
                Object {
                  "$ref": 4,
                },
              ],
              "scope": Object {
                "$ref": 5,
              },
            },
          ],
        },
      ],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 7,
      },
      "variableMap": Object {
        "App": Object {
          "$ref": 1,
        },
        "React": Object {
          "$ref": 0,
        },
      },
      "variableScope": Object {
        "$ref": 6,
      },
      "variables": Array [
        Object {
          "$id": 0,
          "defs": Array [
            Object {
              "name": Object {
                "name": "React",
                "range": Array [
                  12,
                  17,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  7,
                  17,
                ],
                "type": "ImportNamespaceSpecifier",
              },
              "parent": Object {
                "range": Array [
                  0,
                  30,
                ],
                "type": "ImportDeclaration",
              },
              "type": "ImportBinding",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "React",
              "range": Array [
                12,
                17,
              ],
              "type": "Identifier",
            },
          ],
          "name": "React",
          "references": Array [],
          "scope": Object {
            "$ref": 6,
          },
        },
        Object {
          "$id": 1,
          "defs": Array [
            Object {
              "name": Object {
                "name": "App",
                "range": Array [
                  89,
                  92,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  80,
                  164,
                ],
                "type": "FunctionDeclaration",
              },
              "parent": null,
              "type": "FunctionName",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "App",
              "range": Array [
                89,
                92,
              ],
              "type": "Identifier",
            },
          ],
          "name": "App",
          "references": Array [],
          "scope": Object {
            "$ref": 6,
          },
        },
      ],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 7,
  },
  "variables": Array [],
}
`;
